Python/Product/EnvironmentsList/PipExtension.xaml.cs

```diff
@@ -40,30 +40,61 @@ internal sealed partial class PipExtension : UserControl, ICanFocus {
+        private readonly Timer _focusTimer;
 
         public PipExtension(PipExtensionProvider provider) {
             _provider = provider;
             DataContextChanged += PackageExtension_DataContextChanged;
+            _focusTimer = new Timer(FocusWaitExpired);
             InitializeComponent();
         }
 
         void ICanFocus.Focus() {
             Dispatcher.BeginInvoke((Action)(() => {
                 try {
                     Focus();
-                    if (SearchQueryText.IsVisible) {
+                    if (SearchQueryText.IsVisible && SearchQueryText.IsEnabled) {
                         Keyboard.Focus(SearchQueryText);
                     } else {
-                        SearchQueryText.IsVisibleChanged += SearchQueryText_IsVisibleChanged;
+                        // Package manager may still be initializing itself
+                        // Search box is disabled if package manager is not ready
+                        if (!SearchQueryText.IsEnabled) {
+                            SearchQueryText.IsEnabledChanged += SearchQueryText_IsEnabledChanged;
+                        }
+                        if (!SearchQueryText.IsVisible) {
+                            SearchQueryText.IsVisibleChanged += SearchQueryText_IsVisibleChanged;
+                        }
+                        // It may never become ready/enabled, so don't wait for too long.
+                        _focusTimer.Change(1000, Timeout.Infinite);
                     }
                 } catch (Exception ex) when (!ex.IsCriticalException()) {
                 }
             }), DispatcherPriority.Loaded);
         }
 
+        private void FocusWaitExpired(object state) {
+            // Waited long enough, the user might start clicking around soon,
+            // so cancel focus operation.
+            SearchQueryText.IsEnabledChanged -= SearchQueryText_IsEnabledChanged;
+            SearchQueryText.IsVisibleChanged -= SearchQueryText_IsVisibleChanged;
+        }
+
         private void SearchQueryText_IsVisibleChanged(object sender, DependencyPropertyChangedEventArgs e) {
             SearchQueryText.IsVisibleChanged -= SearchQueryText_IsVisibleChanged;
-            Keyboard.Focus(SearchQueryText);
+            if (SearchQueryText.IsVisible && SearchQueryText.IsEnabled) {
+                SearchQueryText.IsEnabledChanged -= SearchQueryText_IsEnabledChanged;
+                Keyboard.Focus(SearchQueryText);
+                _focusTimer.Change(Timeout.Infinite, Timeout.Infinite);
+            }
+        }
+
+        private void SearchQueryText_IsEnabledChanged(object sender, DependencyPropertyChangedEventArgs e) {
+            SearchQueryText.IsEnabledChanged -= SearchQueryText_IsEnabledChanged;
+            if (SearchQueryText.IsVisible && SearchQueryText.IsEnabled) {
+                SearchQueryText.IsVisibleChanged -= SearchQueryText_IsVisibleChanged;
+                Keyboard.Focus(SearchQueryText);
```

Comment:
So previously, UI has been freezing because of the `Keyboard.Focus(SearchQueryText);`?

Reply:
Not freezing, but it wouldn't set focus if it was disabled.
FYI the one place where this is used is when you right click "Install package" from solution explorer. It opens the environments tool window, selects the interpreter, select packages from the drop down combo box, and sets focus to the search box.